### PR TITLE
DIV-4360 Grant user permission if has valid credentials

### PR DESCRIPTION
**Pull request by qzhou-hmcts**
# Description

Grant permission on linking if the user has the right email address.

Fixes # (issue)
DIV-4360

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

**Test Configuration**:

* Hardware:
* O/S and version:
* JDK:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java [hunk 1]
```diff
@@ -173,6 +169,40 @@ public void givenLetterHolderIdAndCaseStateMatches_whenLinkCoRespondent_thenShou
...
+
+        final String respondentFirstName = "respondent-" + UUID.randomUUID().toString();
+
+        final PinResponse pinResponse = idamTestSupport.createPinUser(respondentFirstName);
+
+        Map caseData = ResourceLoader.loadJsonToObject(PAYLOAD_CONTEXT_PATH + "addresses.json", Map.class);
+        caseData.put(CO_RESP_LETTER_HOLDER_ID_FIELD, pinResponse.getUserId());
+
+        Long caseId = ccdClientSupport.submitCase(caseData, getCaseWorkerUser()).getId();
+        System.out.println(caseId);
```
Comment (gurpiarbassi) [inline src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java]: can we use logger?

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java [hunk 2]
```diff
@@ -173,6 +169,40 @@ public void givenLetterHolderIdAndCaseStateMatches_whenLinkCoRespondent_thenShou
         assertEquals(caseId, response.path("id"));
     }
 
+    @Test
+    public void givenLinkedCoRespondent_whenLinkCoRespondent_thenProcessAsNormal()
```
Comment (gurpiarbassi) [inline src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java]: how about: givenLinkedCoRespondent_whenReLinkingCoRespondent_thenProcessAsNormal

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java [hunk 3]
```diff
@@ -35,14 +37,19 @@ public void linkRespondent(String authorisation, String caseId, String letterHol
...
+            throw new CaseNotFoundException(String.format("Case with caseId [%s] and letter holder id [%s] not found",
                 caseId, letterHolderId));
         }
 
         UserDetails respondentUser = getUserDetails(authorisation);
 
+        if (!isValidUserHolder(caseDetails, respondentUser.getEmail(), letterHolderId)) {
+            throw new CaseNotFoundException(String.format("Case with caseId [%s] and letter holder id [%s] "
+                    + "already assigned",
+                caseId, letterHolderId));
```
Comment (hosainnet) [inline src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: It's not really a "`CaseNotFoundException`" in this case is it?

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java [hunk 4]
```diff
@@ -75,7 +82,7 @@ private void grantAccessToCase(UserDetails anonymousCaseWorker, String caseId, S
         );
     }
 
-    private boolean linkingIsValid(CaseDetails caseDetails, String letterHolderId) {
+    private boolean linkingIsValidHolderId(CaseDetails caseDetails, String letterHolderId) {
         if (caseDetails == null || caseDetails.getData() == null || StringUtils.isBlank(letterHolderId)) {
```
Comment (gurpiarbassi) [inline src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: not part of your PR but can we separate the validation of the caseDetails from the letterHolderId?
This method can then be renamed to isValidLetterHolderId()

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java [hunk 5]
```diff
@@ -35,14 +37,19 @@ public void linkRespondent(String authorisation, String caseId, String letterHol
...
-            throw new CaseNotFoundException(String.format("Case with caseId [%s] and letter holder id [%s] not found "
-                    + "or case already has linked respondent",
+        if (!linkingIsValidHolderId(caseDetails, letterHolderId) ) {
+            throw new CaseNotFoundException(String.format("Case with caseId [%s] and letter holder id [%s] not found",
                 caseId, letterHolderId));
         }
 
         UserDetails respondentUser = getUserDetails(authorisation);
 
+        if (!isValidUserHolder(caseDetails, respondentUser.getEmail(), letterHolderId)) {
```
Comment (gurpiarbassi) [inline src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: would isValidUser() be sufficient i.e. get rid of the word 'Holder' ?

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java (hunk 3 above)
Comment (gurpiarbassi) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: Yeh in COS - I created a general purpose ValidationException which I mapped to a BadRequest in the handler. This sounds like the same sort of thing.

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java (hunk 1 above)
Comment (qzhou-hmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java]: Deleted

File: src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java (hunk 2 above)
Comment (qzhou-hmcts) [reply src/integrationTest/java/uk/gov/hmcts/reform/divorce/ccd/LinkRespondentTest.java]: Updated

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java (hunk 4 above)
Comment (qzhou-hmcts) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: I don't see the advantage of create a method to do ' return StringUtils.isBlank(letterHolderId))'

Comment (gurpiarbassi) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: it won't be doing just ```return StringUtils.isBlank(letterHolderId))``` will it? you've still got the bit 
```
return respondentIsValid(caseDetails, letterHolderId)
            || coRespondentIsValid(caseDetails, letterHolderId);
```
afterwards.

alternatively if you want to keep it all in one method, maybe just rename the method to ```isValidCaseAndLetterHolder()``` perhaps?

Comment (qzhou-hmcts) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: do you mean combine `linkingIsValidHolderId` and `isValidUserHolder` ?

Comment (gurpiarbassi) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: no, they can still remain as 2 methods. What I was referring to was the validation of the case and the letterHolderId being done in the same method. I was saying if you don't want to separate it, you could just rename the method to isValidCaseAndLetterHolder()?

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java (hunk 3 above)
Comment (qzhou-hmcts) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: As this changes need to go  ASAP to fix prod bug, I've created this ticket to improve this area.
https://tools.hmcts.net/jira/browse/DIV-4381
Linking respondent needs refactor as we are retrieving the case three times from CCD making this operation really slow.

Comment (gurpiarbassi) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: ok fair enough.

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java (hunk 4 above)
Comment (qzhou-hmcts) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: Renamed to isValidCaseAndLetterHolder

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java (hunk 5 above)
Comment (qzhou-hmcts) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: renamed

File: src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java (hunk 3 above)
Comment (hosainnet) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: Isn't one of the issues here is that CCD is slow and calls sometimes times out? Surely the fact that we're calling them 3 times for 1 call is contributing to this..

Comment (qzhou-hmcts) [reply src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdAccessServiceImpl.java]: CCD calls are itself slow, but if we are calling 3 times to retrieve the same data, it  just make it works, in Spain we call it `add gas oil into fire`. Reduce the number of unneeded calls is something we can improve on our side

**Pull request merged**